### PR TITLE
IGNITE-20869 Got rid of using CompletableFuture#orTimeout method on operations hot path

### PR DESCRIPTION
**Pull request by vldpyatkov**
**FeatureTimeoutBenchmark**
```
*  Results on 11th Gen Intel® Core™ i7-1165G7 @ 2.80GHz, openjdk 11.0.24, Windows 10 Pro:
 * Benchmark                     (useFutureEmbeddedTimeout)  Mode  Cnt   Score    Error  Units
 * FeatureTimeoutBenchmark.test                       false  avgt   20   1,501 ±  0,058  us/op
 * FeatureTimeoutBenchmark.test                        true  avgt   20  32,573 ± 47,598  us/op
```

https://issues.apache.org/jira/browse/IGNITE-20869

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+ * Feature timeout benchmark.
+ */
+@State(Scope.Benchmark)
+@Fork(1)
+@Threads(1)
+@Warmup(iterations = 10, time = 2)
+@Measurement(iterations = 20, time = 2)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+public class FeatureTimeoutBenchmark {
```
Comment (rpuch) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java:70]: What does 'feature' mean here? Is it just 'future'?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 2]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
+        /** Worker sleep interval. */
+        private final long sleepInterval = getLong("IGNITE_TIMEOUT_WORKER_SLEEP_INTERVAL", 500);
+
+        /** Active operation. */
+        public final ConcurrentMap<Long, TimeoutObject> requestsMap;
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:815]: Why is it declared as `ConcurrentMap` and not just `Map`?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 3]
```diff
@@ -87,7 +97,7 @@ class TcpClientChannel implements ClientChannel, ClientMessageHandler, ClientCon
     private final AtomicLong reqId = new AtomicLong(1);
 
     /** Pending requests. */
-    private final Map<Long, ClientRequestFuture<?>> pendingReqs = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Long, TimeoutObject> pendingReqs = new ConcurrentHashMap<>();
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:100]: Do we really need this change from `Map`?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 4]
```diff
@@ -104,6 +114,9 @@ class TcpClientChannel implements ClientChannel, ClientMessageHandler, ClientCon
     /** Executor for async operation listeners. */
     private final Executor asyncContinuationExecutor;
 
+    /** Timeout worker. */
+    private TimeoutWorker timeoutWorker;
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:118]:
```suggestion
    private final TimeoutWorker timeoutWorker;
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 5]
```diff
@@ -158,6 +178,8 @@ private CompletableFuture<ClientChannel> initAsync(ClientConnectionMultiplexer c
                         log.debug("Connection established [remoteAddress=" + s.remoteAddress() + ']');
                     }
 
+                    new IgniteThread(timeoutWorker).start();
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: Does this mean a thread per client? Isn't this too much?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 6]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+
+        @Override
+        protected void body() {
+            try {
+                TimeoutObject timeoutObject;
+
+                while (!isCancelled()) {
+                    long now = coarseCurrentTimeMillis();
+
+                    for (Entry<Long, TimeoutObject> entry : new HashMap<>(requestsMap).entrySet()) {
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:845]: Why is a copy of the map created here? Isn't concurrent map safe with respect to iteration happening concurrently to modifications?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 7]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+                        updateHeartbeat();
+
+                        timeoutObject = entry.getValue();
+
+                        assert timeoutObject != null : "Unexpected null on the timeout queue.";
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:850]: The map is not sorted, so why is it a queue?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 8]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+
+                        if (timeoutObject.getEndTime() > 0 && now > timeoutObject.getEndTime()) {
+                            CompletableFuture<?> fut = timeoutObject.getFuture();
+
+                            if (!fut.isDone()) {
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:855]: Is this check really needed? If the future is already completed, its subsequent completion will have no effect, and it will take just one volatile read (which we pay for `isDone()`)

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 9]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+                                fut.completeExceptionally(new TimeoutException());
+                            }
+
+                            // Client-facing future will fail with a timeout, but internal ClientRequestFuture will stay in the map -
+                            // otherwise we'll fail with "protocol breakdown" error when a late response arrives from the server.
+                            // requestsMap.remove(entry.getKey(), timeoutObject)
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:861]: Why is this line commented out?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 10]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+                        Thread.sleep(sleepInterval);
+                    } catch (InterruptedException e) {
+                        log.info("The timeout worker was interrupted, probably the client is stopping.");
+                    }
+
+                    updateHeartbeat();
+                }
+
+            } catch (Throwable t) {
+                // Call FH here.
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:875]: If we are going to call a FH, let's add a TODO for this

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 11]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+            }
+        }
+    }
+
+    /**
+     * Timeout object.
+     * The class is a wrapper over the complete future.
+     */
+    private static class TimeoutObject {
+        /** End time. */
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:887]:
```suggestion
        /** End time (milliseconds since Unix epoch). */
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 12]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+            this.endTime = endTime;
+            this.fut = fut;
+        }
+
+        /**
+         * Gets end timestamp.
+         *
+         * @return End timestamp in milliseconds.
+         */
+        public long getEndTime() {
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:909]:
```suggestion
        public long endTime() {
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 13]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
...
+            return endTime;
+        }
+
+        /**
+         * Gets a target future.
+         *
+         * @return A future.
+         */
+        public ClientRequestFuture<?> getFuture() {
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:918]:
```suggestion
        public ClientRequestFuture<?> future() {
```

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java [hunk 14]
```diff
@@ -287,10 +302,9 @@ private CompletableFuture<NetworkMessage> invoke0(ClusterNode recipient, Channel
 
         long correlationId = createCorrelationId();
 
-        CompletableFuture<NetworkMessage> responseFuture = new CompletableFuture<NetworkMessage>()
-                .orTimeout(timeout, TimeUnit.MILLISECONDS);
+        CompletableFuture<NetworkMessage> responseFuture = new CompletableFuture<>();
 
-        requestsMap.put(correlationId, responseFuture);
+        requestsMap.put(correlationId, new TimeoutObject(timeout > 0 ? coarseCurrentTimeMillis() + timeout : 0, responseFuture));
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:307]: It seems that `orTimeout(0, SECONDS)` will timeout immediately, but the `TimeoutObject` treats 0 as 'no timeout at all', so this changes the behavior. I think we need to establish what we want here first.

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java [hunk 15]
```diff
@@ -764,4 +782,116 @@ private InetSocketAddress getFromCacheOrCreateResolved(ClusterNode recipientNode
...
     }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
+        /** Worker sleep interval. */
+        private final long sleepInterval = getLong("IGNITE_TIMEOUT_WORKER_SLEEP_INTERVAL", 500);
+
+        /** Active operation. */
```
Comment (rpuch) [inline modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:793]:
```suggestion
        /** Active operations. */
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 16]
```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
             }
         }
     }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:810]: `TimeoutWorker` classes seem almost identical and `TimeoutObject` classes only differ in the element type of the future. It seems that it's really easy to have just one generic `TimeoutObject`. As for `TimeoutWorker`, most its logic is identical, and differences can be extracted to an abstract method, to avoid code duplication.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java [hunk 17]
```diff
@@ -87,23 +87,22 @@ public ClusterService clusterService() {
     ) {
         CompletableFuture<Message> fut = new CompletableFuture<>();
 
-        fut.orTimeout(timeoutMs, TimeUnit.MILLISECONDS).
-            whenComplete((res, err) -> {
-                assert !(res == null && err == null) : res + " " + err;
+        fut.whenComplete((res, err) -> {
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java:90]: This change seems to change the semantics. Before the change, `whenComplete()` callback would see the `TimeoutException`, but now it won't.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java [hunk 18]
```diff
@@ -121,6 +120,10 @@ else if (err instanceof TimeoutException) // Translate timeout exception.
 
                 blockedMsgs.add(msgData);
 
+                if (timeoutMs > 0) {
+                    fut.orTimeout(timeoutMs, TimeUnit.MILLISECONDS);
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java:124]: Could you please elaborate why this is better? We still call `orTimeout()`, just later.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+    public void test() {
+        if (useFutureEmbeddedTimeout) {
+            for (int i = 0; i < 10; i++) {
+                var fut = new CompletableFuture<Void>();
+
+                futs.put(ID_GEN.incrementAndGet(), fut);
+
+                fut.orTimeout(10, TimeUnit.MILLISECONDS);
+
+                if (futs.size() > 100_000) {
```
Comment (rpuch) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java]: Why is this needed?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+        } else {
+            assert waitForCondition(requestsMap::isEmpty, 10_000);
+
+            timeoutWorker.interrupt();
+            timeoutWorker = null;
+        }
+    }
+
+    /**
+     * Benchmark for KV upsert via embedded client.
```
Comment (rpuch) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java:129]: It's not about KV at all :)

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+                .build();
+
+        new Runner(opt).run();
+    }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutRunnable implements Runnable {
+        /** Active operation. */
```
Comment (rpuch) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java:174]:
```suggestion
        /** Active operations. */
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 3 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:100]: Because the way that we use the map requires it has to be concurrent.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 2 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:815]: If someone uses a plain map here, they get a compile error.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: Do you think we can share the worker among clients started on one JVM?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:845]: This approach is better than that where we use a concurrent iterator.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 3 above)
Comment (rpuch) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:100]: Why? :)

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (rpuch) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: Why not :) The collection is concurrent anyway

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 8 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:855]: But we would have to create exception instances. This code is not in a low-latency path, and I was worried about GC here.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java (hunk 17 above)
Comment (vldpyatkov) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java:90]: We still set the timeout on sending the operation.
It just has excessive usage of `orTimeoput`, that I removbed.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java (hunk 18 above)
Comment (vldpyatkov) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/rpc/impl/IgniteRpcClient.java:124]: It is a code for tests in which we block RAFT commands.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java (hunk 19 above)
Comment (vldpyatkov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/FeatureTimeoutBenchmark.java]: I avoided GC here.

**Comment by ptupitsyn** [comment]
@vldpyatkov could you please try running `ClientHeartbeatBenchmark` with the following modification to enable timeouts:

```
         client = (TcpIgniteClient) IgniteClient.builder()
                 .addresses("127.0.0.1:" + testServer.port())
+                .operationTimeout(5000)
                 .build();
```

On my machine, I don't see any difference between main and your branch.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: One timeout worker for all clients is enough. Maybe even one timeout worker per JVM for all Ignite needs, we can put it in the core module.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 6 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:845]: We iterate over the map to copy it anyway, it can't be faster than just iterating. Something must be wrong with the benchmark.

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java [hunk 22]
```diff
@@ -764,4 +782,116 @@ private InetSocketAddress getFromCacheOrCreateResolved(ClusterNode recipientNode
 
         return address != null ? address : createResolved(recipientNode.address());
     }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
```
Comment (ptupitsyn) [inline modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:789]: Duplicate code. As mentioned above, we can put this into core module and reuse.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 7 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:850]: Fixed it.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 9 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:861]: Because of that comment above,. the protocol has to get a response in any case. It is a main difference between timeout worker for Ignite node and the client. That is also a reason why I do not use one class in both cases.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: I do not know how I can do it right now.
May I left TODO here?

Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: Do you have any questions, what exactly is not clear?
* Create a separate class
* Put a static concurrent map there
* Create one thread for timeout checks

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:845]: I tested it again and did not find any difference. I will use the map iterator.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:181]: It is not clear where this thread has to be stopped.
A static variable does not seem like a good solution to me.
I believe it is better to do it on a separate ticket.

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java (hunk 22 above)
Comment (vldpyatkov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java]: I moved the code in one palce.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 10 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java]: Added TODO

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java (hunk 14 above)
Comment (vldpyatkov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:303]: I took this code from our repository. It was for a thin client.
Anyway, timeout 0 is something specific that will be used in a test.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 16 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java]: I moved this class to another place, where it is available for both usages.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (rpuch) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:182]: Is a demon thread an option?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 9 above)
Comment (rpuch) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java]: Then why do you leave the commented-out line? You explained why you don't remove from the map, isn't this enough?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 23]
```diff
@@ -395,7 +419,7 @@ private <T> void complete(ClientRequestFuture<T> pendingReq, ClientMessageUnpack
     private void processNextMessage(ClientMessageUnpacker unpacker) throws IgniteException {
         if (protocolCtx == null) {
             // Process handshake.
-            complete(pendingReqs.remove(-1L), unpacker);
+            complete((ClientRequestFuture<?>) pendingReqs.remove(-1L).future(), unpacker);
```
Comment (rpuch) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:422]: Let's make `TimeoutObject` generic with `T` being the type inside the `CompletableFuture`. This will allow to avoid these casts.

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java (hunk 14 above)
Comment (ptupitsyn) [reply modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:303]: Thin client treats 0 as "no timeout" (which is documented in the config javadoc), this might be different in other places.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 5 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:182]: Right now, we start/stop it in the general life cycle. I do not see a reason to make it a demon.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 9 above)
Comment (vldpyatkov) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java]: I have rewritten this place.

File: modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java (hunk 14 above)
Comment (vldpyatkov) [reply modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java:303]: That is exactly why I wrote it in this way (TimeoutWorker#body).

**Review by rpuch — APPROVED** [review]

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 24]
```diff
@@ -314,13 +335,16 @@ private <T> ClientRequestFuture<T> send(
...
+            long timeout
+    ) {
         if (closed()) {
             throw new IgniteClientConnectionException(CONNECTION_ERR, "Channel is closed", endpoint());
         }
 
         ClientRequestFuture<T> fut = new ClientRequestFuture<>(payloadReader, notificationFut);
-        pendingReqs.put(id, fut);
+
+        pendingReqs.put(id, new TimeoutObject<>(timeout > 0 ? coarseCurrentTimeMillis() + timeout : 0, fut));
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:347]: Can we make `TimeoutObject` an interface? We already use a custom future class (`ClientRequestFuture`), it can hold the timeout as well, so we avoid an allocation of `TimeoutObject` wrapper.

**Review by ptupitsyn — APPROVED** [review]
